### ISSUE TITLE
Support for Huawei Contact Shield

### ISSUE THREAD

**Issue by abusuioc** [DP-3T/dp3t-sdk-android#179]
Let me start by thanking the whole team for the great job you guys did on this SDK.

As you well know, because of the [US trade ban](https://en.wikipedia.org/wiki/China%E2%80%93United_States_trade_war) , the new and upcoming Huawei devices powered by Android cannot anymore use Google Play Services (GMS). As a response to that, Huawei developed its own alternative to GMS - called Huawei Mobile Services (HMS). As part of the HMS offering, [Contact Shield](https://developer.huawei.com/consumer/en/doc/Contact-Shield-V1/introduction-0000001050738511-V1) offers the same capabilities as the GMS equivalent that you already depend on: Exposure Notifications.

It would be fantastic if the SDK would offer support for HMS Contact Shield, in addition to the current support for GMS Exposure Notifications. Huawei users wish to contribute to the global effort of tracing the spread of Covid-19 and, currently, they cannot -  since the DP-3T SDK and the apps built on top of it require the presence of GMS.

**Comment by marado** [comment]
FYI, #143 shows how can the prestandard tag be used to support non-GMS devices.

I know that isn't the same as supporting Contact Shield, and a better solution would be having the same app using GMS, HMS or nothing, depending on what is available on the system, but this information might still be useful.

**Comment by abusuioc** [comment]
Thanks for pointing that out, @marado !

However, the ideal scenario is exactly what you proposed: "having the same app using GMS, HMS or nothing, depending on what is available on the system".

But that's not the case at the moment. For instance, what the Swiss Covid app does right now is to block further access if GMS is not available on the device. This focus only on GMS leaves a gap for non-GMS devices - which are usually powered by HMS.

At a first look, Contact Shield API looks very similar to Exposure Notifications - it would be nice to get additional feedback from the engineers working on the SDK since they know the implementation needs inside out.

Are there plans to support HMS devices?

**Comment by ineiti** [comment]
@abusuioc will the HMS/Contact Shield be BLE compatible with the GAEN solution? And do you know if there is source code available?

It would be interesting to have a GAEN-API compatible library in Open Source form to enable services for Huawei, or other non-GMS enabled devices.

**Comment by abusuioc** [comment]
Yes @ineiti , afaik HMS/ContactShield (CS) is compatible with GAEN in terms of BLE usage and API def.
I could found some [development samples](https://developer.huawei.com/consumer/en/doc/Contact-Shield-V1/contact-procedure-0000001050698482-V1) and a [full API overview](https://developer.huawei.com/consumer/en/doc/Contact-Shield-V1/contactshield-overview-0000001051058454-V1) .

It would be nice if some of the engineers working on the dp3t-sdk-android could take and look at the above links and judge if CS offers all they need to have dp3t-sdk-android running on top of CS.
